emark/backends.py

```diff
@@ -1,60 +1,78 @@
-from django.core.mail.message import sanitize_address
 
 from emark import models
 from emark.message import MarkdownEmail
 
 __all__ = [
     "ConsoleEmailBackend",
+    "SMTPEmailBackend",
     "TrackingConsoleEmailBackend",
     "TrackingSMTPEmailBackend",
 ]
 
 
+class RenderEmailBackendMixin:
+    def send_messages(self, email_messages):
+        self._messages_sent = []
+        for message in email_messages:
+            if isinstance(message, MarkdownEmail):
+                message.render()
+        try:
+            return super().send_messages(email_messages)
+        finally:
+            models.Send.objects.bulk_create(self._messages_sent)
+
+
 class TrackingEmailBackendMixin:
     """Add tracking framework to an email backend."""
 
     def send_messages(self, email_messages):
         self._messages_sent = []
+        for message in email_messages:
+            if isinstance(message, MarkdownEmail):
+                message.render(tracking_uuid=uuid.uuid4())
         try:
             return super().send_messages(email_messages)
         finally:
             models.Send.objects.bulk_create(self._messages_sent)
 
-    def _track_message_clone(self, clone, message):
-        if isinstance(clone, MarkdownEmail):
+    def _track_message(self, message: EmailMessage):
+        if isinstance(message, MarkdownEmail):
             self._messages_sent.append(
                 models.Send(
-                    pk=clone._tracking_uuid,
-                    from_address=message["From"],
-                    to_address=message["To"],
-                    subject=message["Subject"],
-                    body=clone.body,
-                    html=clone.html,
-                    user=getattr(clone, "user", None),
-                    utm=clone.get_utm_params(**clone.utm_params),
+                    pk=message.uuid,
+                    from_email=message.from_email,
+                    to=message.to,
+                    cc=message.cc,
+                    bcc=message.bcc,
+                    reply_to=message.reply_to,
+                    subject=message.subject,
+                    body=message.body,
+                    html=message.html,
+                    user=getattr(message, "user", None),
+                    utm=message.get_utm_params(),
                 )
             )
         else:
             self._messages_sent.append(
                 models.Send(
-                    pk=clone._tracking_uuid,
-                    from_address=message["From"],
-                    to_address=message["To"],
-                    subject=message["Subject"],
-                    body=clone.body,
+                    from_email=message.from_email,
+                    to=message.to,
+                    cc=message.cc,
+                    bcc=message.bcc,
+                    reply_to=message.reply_to,
+                    subject=message.subject,
+                    body=message.body,
```

Comment:
Why don't we want to save `html` in case it is available (when it is a Django `EmailMessage`, not an emark one)?

Reply:
Cause there is no `html`-attribute ;)

Reply:
Lol, true, sorry :)